### PR TITLE
#169817543 Reset Password

### PR DESCRIPTION
**Pull request by erickyvand**
#### What does this PR do?
- Has to reset password in case the user forgot password
#### Description of Task to be completed?
- Has the endpoint of found user and send email working and reset password working
#### How should this be manually tested?
- using postman  when user got token from email should include it in the header then access the reset password endpoint
#### What are the relevant pivotal tracker stories?
- [#169817543](https://www.pivotaltracker.com/story/show/169817543)
#### Screenshots 
Screen Shot 2019-12-27 at 09 53 37

### CODE REVIEW

File: src/controllers/UserController.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      res.send('No user found');
+    } else {
+      jwt.sign({ user }, 'secret', (err, token) => {
+        const transporter = nodemailer.createTransport({
+          host: 'smtp.gmail.com',
+          port: 587,
+          secure: false,
+          auth: {
+            user: 'icyiiddy@gmail.com',
+            pass: '123sultan'
```
Comment (victor-abz) [inline src/controllers/UserController.js]: Don't hardcode the values. you can use env variables.

File: src/services/UserService.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    });
+    return user;
+  }
+
+  static async resetPassword(email, password) {
+    // const userUpdate = await database.users.findOne({
+    //   where: {
+    //     email
+    //   }
+    // });
```
Comment (victor-abz) [inline src/services/UserService.js]: Remove unnecessary comments

File: src/models/index.js [hunk 3]
```diff
@@ -1,19 +1,20 @@
...
+const environment = {
+  development,
+  production,
+  test
+};
 
 const basename = path.basename(__filename);
 const env = process.env.NODE_ENV || 'development';
-const config = dbConfig[env];
+const config = environment[env];
```
Comment (victor-abz) [inline src/models/index.js]: why did you change here?
I think to maintain the way the file was we can do`
```
import { development, test, production } from '../config/config';
const dbConfig = { development, test, production };
```
```suggestion
const config = dbConfig[env];
```

File: src/controllers/reset.controller.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  /**
+  * @param {Request} req
+  * @param {Response} res
+  * @returns {FindUser} the function find if user exist in database
+  */
+  static async findUser(req, res) {
+    const { email } = req.body;
+    const user = await ResetService.findUser(email);
+
+    if (!user) {
```
Comment (gadishimwe) [inline src/controllers/reset.controller.js]: @erickyvand , you need to move all validations into middleware file. only controller must handle valid data

File: src/controllers/reset.controller.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  * @param {Request} req
+  * @param {Response} res
+  * @returns {resetPassword} the function reset a password of a user
+  */
+  static async resetPassword(req, res) {
+    const { email, newPassword, confirmPass } = req.body;
+    const findUser = await ResetService.findUser(email);
+
+    jwt.verify(req.token, process.env.SECRET_KEY, (err, auth) => {
+      if (err) {
```
Comment (gadishimwe) [inline src/controllers/reset.controller.js]: move also this to middleware file

**Review by gadishimwe — CHANGES_REQUESTED** [review]

File: src/controllers/reset.controller.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  /**
+  * @param {Request} req
+  * @param {Response} res
+  * @returns {resetPassword} the function reset a password of a user
+  */
+  static async resetPassword(req, res) {
+    const { email, newPassword, confirmPass } = req.body;
+    const findUser = await ResetService.findUser(email);
+
+    jwt.verify(req.token, process.env.SECRET_KEY, (err, auth) => {
```
Comment (gadishimwe) [inline src/controllers/reset.controller.js]: also move this too

File: src/controllers/auth.controller.js [hunk 7]
```diff
@@ -84,6 +85,36 @@ class AuthController {
     ResponseService.setSuccess(200, 'Successful logout');
     return ResponseService.send(res);
   }
+
+  /**
+    * @param {Request} req
+    * @param {Response} res
+    * @returns {FindUser} this function finds if user exist in database
+    */
+  static async findUser(req, res) {
```
Comment (victor-abz) [inline src/controllers/auth.controller.js]: I think that this function name is misleading. It finds a user but also send an email to him. so find user might relate to just fetching from DB. Why not call it another name like `sendEmail`

File: src/server.js [hunk 8]
```diff
@@ -1,4 +1,5 @@
 import registerApiDocs from './swagger';
+
```
Comment (victor-abz) [inline src/server.js:3]: Why this changes here

Comment (erickyvand) [reply src/server.js:3]: Nothing changed I did rebase Develop, this file is the same as from Develop

**Review by gadishimwe — APPROVED** [review]

File: src/tests/auth/find.test.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
+import chai from 'chai';
```
Comment (higustave-ops) [inline src/tests/auth/find.test.js]: better you rename it to `<find-user.test.js>` because you are finding a user. `<find.test.js>` is not descriptive

**Review by victor-abz — APPROVED** [review]
LGTM

**Comment by katunold** [comment]
@erickyvand could you kindly ensure that the response object is in JSON format

**Review by katunold — COMMENTED** [review]
@erickyvand kindly reword your commit message to have the standard form of header, body and footer.

**Review by katunold — APPROVED** [review]
LGTM
